lib/components/user/places/favorite-place-page.js

```diff
@@ -0,0 +1,210 @@
+import { withAuthenticationRequired } from '@auth0/auth0-react'
+import clone from 'clone'
+import { Form, Formik } from 'formik'
+import React, { Component } from 'react'
+import { Button } from 'react-bootstrap'
+import { connect } from 'react-redux'
+import styled from 'styled-components'
+import * as yup from 'yup'
+
+import * as uiActions from '../../../actions/ui'
+import * as userActions from '../../../actions/user'
+import DesktopNav from '../../app/desktop-nav'
+import Icon from '../../narrative/icon'
+import { RETURN_TO_CURRENT_ROUTE } from '../../../util/ui'
+import { cloneForEditing, isHomeOrWork, PLACE_TYPES } from '../../../util/user'
+import withLoggedInUserSupport from '../with-logged-in-user-support'
+import PlaceEditor from './place-editor'
+
+// Styled components
+const SaveButton = styled(Button)`
+  float: right;
+`
+const BackButton = styled(Button)`
+  padding: 0;
+`
+
+// Custom places without home or work
+const customPlaces = clone(PLACE_TYPES)
+delete customPlaces.home
+delete customPlaces.work
+
+const BLANK_PLACE = {
+  ...PLACE_TYPES.custom,
+  address: '',
+  name: ''
+}
+
+// The form fields to validate.
+const validationSchemaShape = {
+  address: yup.string().required('Please set a location for this place'),
+  name: yup.string().required('Please enter a name for this place')
+}
+
+/**
+ * Lets the user edit the details (address, type, nickname)
+ * of a new or existing favorite place.
+ */
+class FavoritePlacePage extends Component {
+  /**
+   * Navigates back to the previous browser location,
+   * whether it is the accounts page or another page that linked to the current URL.
+   */
+  // TODO: Refine this behavior.
+  _handleBack = () => {
+    window.history.back()
+  }
+
+  /**
+   * Silently save the changes to the loggedInUser, and go to the previous URL.
+   */
+  _handleSave = async placeToSave => {
+    // Update the icon for the place type.
+    placeToSave.icon = PLACE_TYPES[placeToSave.type].icon
+
+    // Save changes to loggedInUser
+    const { createOrUpdateUser, loggedInUser, match } = this.props
+
+    // Get the places as shown (and not as retrieved from db), so that the index passed from URL applies
+    // (indexes 0 and 1 are for 'Home' and 'Work' places).
+    const clonedUser = cloneForEditing(loggedInUser)
+    const placeIndex = match.params.id
+
+    if (placeIndex === 'new') {
+      clonedUser.savedLocations.push(placeToSave)
+    } else {
+      clonedUser.savedLocations[placeIndex] = placeToSave
+    }
+
+    await createOrUpdateUser(clonedUser, true)
+
+    this._handleBack()
+  }
```

Comment:
Most of this logic seems like it would be better placed inside an action.

Reply:
Good point. Updated in https://github.com/opentripplanner/otp-react-redux/pull/280/commits/53797da17e0d91c1c07659bbd169f4e14ec8683f.